### PR TITLE
chore(docs): use babel-loader

### LINKED ISSUE #508: RFC: switch to Babel
_Previously we've decided to switch to Babel as compiler to improve our developer experience and get benefits from Babel's ecosystem like plugins. I've separated our switch to small steps that will allow us to make this process clearly._

## 1. Switch docs

I think that our main issue now is very slow work of `yarn start` i.e. start docs and we should solve this problem firstly.

### Actions

- [x] compare the performance of `babel-loader` & `fork-ts-checker-webpack-plugin` VS `awesome-typescript-loader` with `useBabel=true`

`awesome-typescript-loader` doesn't work properly with Babel 7, consider use `babel-loader` because it faster. Requires changes in project structure as we can't use required `--
-isolatedModules` to work properly: https://github.com/babel/babel/issues/8361

### Goals

- docs will start faster (_performance_)
- errors in types should not break webpack's process: _it's really painful when everything is done and it fails on the final step._
- hot reload will work properly https://github.com/gaearon/react-hot-loader#typescript

## 2. Switch to `@babel/register`

The second step is the replacement of `ts-node` with `@babel/register` this will improve performance of all our internal scripts.

## 3. Switch to `gulp-babel`

Replace `gulp-typescript` with `gulp-babel` or even with `pika-pkg`: #1169.

### Use existing plugins

`babel-plugin-lodash` & `babel-plugin-transform-react-remove-prop-types` will allow us to reduce bundle size for consumers.

### PR DESCRIPTION
**Pull request by layershifter**
Related #508.

### Enable `isolatedModules` in TS `compilerOptions`

Babel behaves as if the `--isolatedModules` option was passed to the TypeScript Compiler and this can't be worked around because Babel doesn't support cross-file analysis.

How does Babel handle TypeScript code? **It removes it**, it strips out all the TypeScript, turns it into "regular" JavaScript. Any bundler don't know anything about types because don't exists in runtime and that's why types can't be reexported.

https://babeljs.io/docs/en/babel-plugin-transform-typescript

### Updates in exports

As mentioned before we can't reexport types, that's why exports were transformed:

#### Before
```ts
export { default as Accordion, AccordionProps } from './Accordion'
```
#### After
```ts
export { default as Accordion } from './Accordion'
export * from './Accordion' // will AccordionProps and everything except default
```

#### `types.internal.ts`

These files will contain shared and reusable types in package, but these types shouldn't be part of a public API and can't be exported.

```js
export * from './types' // contains only types that are part of public API
```

### Remove `awesome-typescript-loader` & `ts-loader`

Was replaced with `babel-loader` and `fork-ts-checker-webpack-plugin` to keep type checks.

### Add `fork-ts-checker-webpack-plugin`

Webpack plugin that runs TypeScript type checker on a separate process. It allows very fast compile code by `babel-loader` and run type checks parallel, so by switching to Babel we will not loose TS checks.

---

_This PR do not resolve issues with HMR._

### CODE REVIEW

File: packages/react/src/components/Loader/Loader.tsx [hunk 1]
```diff
@@ -142,6 +142,7 @@ class Loader extends UIComponent<WithAsProp<LoaderProps>, LoaderState> {
           {Box.create(label, {
             defaultProps: { className: Loader.slotClassNames.label, styles: styles.label },
           })}
+          222
```
Comment (miroslavstastny) [inline packages/react/src/components/Loader/Loader.tsx]: ?

Comment (layershifter) [reply packages/react/src/components/Loader/Loader.tsx]: Cruft, removed it :)

File: build/tsconfig.commonjs.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "./tsconfig.common.json",
   "compilerOptions": {
+    "isolatedModules": false,
```
Comment (kuzhelov) [inline build/tsconfig.commonjs.json:4]: this config and the next one are used for `yarn build` scenarios, where Babel is not used.

This is necessary to ensure that type declarations will be generated.

File: docs/src/components/ComponentDoc/ComponentExample/index.tsx [hunk 3]
```diff
@@ -1 +1,2 @@
-export { default, ComponentExampleProps } from './ComponentExample'
+export * from './ComponentExample'
```
Comment (kuzhelov) [inline docs/src/components/ComponentDoc/ComponentExample/index.tsx:1]: these changes in exports are necessary to make TS with `--isolatedModules` flag work - this flag ensures that TS compiler uses the same 'file-by-file' processing strategy as Babel

File: packages/react/src/types.ts [hunk 4]
```diff
@@ -8,13 +8,8 @@ export type Extendable<T, V = any> = T & {
   [key: string]: V
 }
 
-export type Nullable<T> = T | null
```
Comment (kuzhelov) [inline packages/react/src/types.ts:11]: all these are unused, right?

File: perf/src/index.tsx [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import '@babel/polyfill'
```
Comment (kuzhelov) [inline perf/src/index.tsx:1]: could you, please, suggest why this change was necessary? where it is used?

**Comment by kuzhelov** [comment]
just couple of moments that we might add to the description (all those are mentioned in the comments):
- brief explanation on how transpilation logic has changed - from recursive modules resolution used by Typescript to isolated files/modules used by Babel. This will serve as concise reasoning behind `--isolatedModules` flag that is introduced to TS compiler's config
- provide brief explanation on why export statements need rework
- provide explanation on the strategy with `types.private` (mb `types.internal`, just to better reflect semantics?) modules - whey these are necessary and which problem they help to address

**Review by kuzhelov — APPROVED** [review]

File: perf/src/index.tsx (hunk 5 above)
Comment (layershifter) [reply perf/src/index.tsx:1]: It contains `regeneratorRuntime` that it's required by `async/await` transform.

File: packages/react/src/types.ts (hunk 4 above)
Comment (layershifter) [reply packages/react/src/types.ts:11]: Exactly 😇
